lib/proxy/router.go

```diff
@@ -280,6 +363,53 @@ func (r *Router) DialHost(ctx context.Context, from net.Addr, host, port, cluste
 	return newProxiedMetricConn(conn), trace.Wrap(err)
 }
 
+func preformRBACCheck(ctx context.Context, idInfo IdentityInfo, cluster, targetCluster string, client auth.ClientI, target types.Server) error {
+	// we don't need to check the RBAC for the node if they are only allowed to join sessions
+	if idInfo.Login == teleport.SSHSessionJoinPrincipal && auth.RoleSupportsModeratedSessions(idInfo.AccessChecker.Roles()) {
+		return nil
+	}
+
+	authPref, err := client.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	state := idInfo.AccessChecker.GetAccessState(authPref)
+	_, state.MFAVerified = idInfo.Certificate.Extensions[teleport.CertExtensionMFAVerified]
+	state.EnableDeviceVerification = true
+	state.DeviceVerified = authz.IsSSHDeviceVerified(idInfo.Certificate)
+
+	// check if roles allow access to server
+	if err := idInfo.AccessChecker.CheckAccess(
+		target,
+		state,
+		services.NewLoginMatcher(idInfo.Login),
+	); err != nil {
+		return trace.AccessDenied("user %s@%s is not authorized to login as %v@%s: %v",
+			idInfo.TeleportUser, cluster, idInfo.Login, targetCluster, err)
+	}
+
+	return nil
+}
+
+func eventDeviceMetadataFromCert(cert *ssh.Certificate) *apievents.DeviceMetadata {
```

Comment:
This function is also copied from `lib/srv`. If you want to reuse it, move the function to a common place where it can be reused in all required places.